contracts/utils/OptionsPurchaseQueue.sol

```diff
@@ -0,0 +1,442 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.4;
+
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeMath} from "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {
+    SafeERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {IOptionsPurchaseQueue} from "../interfaces/IOptionsPurchaseQueue.sol";
+import {IRibbonThetaVault} from "../interfaces/IRibbonThetaVault.sol";
+import {IGnosisAuction} from "../interfaces/IGnosisAuction.sol";
+import {Vault} from "../libraries/Vault.sol";
+
+contract OptionsPurchaseQueue is Ownable, IOptionsPurchaseQueue {
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    /**
+     * @dev Contains purchase request info
+     * @param optionsAmount Amount of options to purchase
+     * @param premiums Total premiums the buyer is spending to purchase the options (optionsAmount * ceilingPrice)
+     *  We need to track the premiums here since the ceilingPrice could change between the time the purchase was
+     *  requested and when the options are sold
+     * @param buyer The buyer requesting this purchase
+     */
+    struct Purchase {
+        uint128 optionsAmount; // Slot 0
+        uint128 premiums;
+        address buyer; // Slot 1
+    }
+
+    /// @notice Gnosis Easy Auction
+    IGnosisAuction public immutable gnosisEasyAuction;
+
+    /************************************************
+     *  STORAGE
+     ***********************************************/
+
+    /// @notice Stores the purchase queue for each vault
+    mapping(address => Purchase[]) public purchases;
+
+    /// @notice Stores the total options being purchased for each vault
+    mapping(address => uint256) public override totalOptionsAmount;
+
+    /// @notice Stores the amount of options the vault is allocating towards the purchase queue
+    /// @dev When vaultAllocatedOptions != 0, new purchase requests for the vault are blocked since the vault has
```

Comment:
nit, do you mean `When vaultAllocatedOptions == 0`?

Reply:
New purchase requests get blocked when `vaultAllocatedOptions != 0`, see: https://github.com/ribbon-finance/ribbon-v2/pull/241/files#diff-6d059eef33889a02d96f146813863a0b1a958de6e3730d8ef14ecdc88f4db0e3R144-R145